extended ID fix

WilliamLim781 opened the pull request:
fixed the filter config for extended and changed the callback function to a flag

### Summary
fixed the filter id config and made the callback function into a flag based interrupt
### Remarks

### Checks
- [x] Tested
- [x] Stakeholder approval

WilliamLim781 committed:
extended ID fix

fixed the filter config for extended and changed the callback function to a flag

WilliamLim781 committed:
Merge branch 'develop' into feature/20_Throttle-Position-Sensor

WilliamLim781 committed:
multiple filters, parsing and decoding, and refactoring code

This commit is comprised of three major parts
	1.) Multiple filters enabled
	2.) parsing function
	3.) Decoding function

Multiple filters
	including multiple filters is rather straightforward. All you need to do is reinitialize the CAN filter structure but change the filter bank you are using. To do this is rather simple and I could have done it manually for all IDs but It would have taken a lot of space. So, in order to make the code easier I went ahead and made a function that takes in two different parameters. The first parameter is filter bank number and the second is the extended ID you want to filter. I also made a variable that is called CANIDList which is an array that holds all the IDs.I made this to make to make CAN filtering a little easier on the eyes.

Parsing function
	this function sorts the filtered extended IDs that have passed through the CAN controller. In addition, the function calls another function called canDecodeDataFrame(). This function converts the data frame from each message into either an unsigned integer or a signed integer ( this just depends on the data that is being decoded). After the data is decoded, it is then multiplied by its resolution value  ( you can also divide if you like fractions better). I didn't see any trend in how the resolution was ordered so I multiplied the decoded data manually.

Decoding function
	This is arguably the most important part of this commit. This function utilizes the equation given in the application note to decode all of the data into either an unsigned integer or a signed integer. This function was made to automate the decoding process because I noticed that there were 5 distinct ways the data was stored in a single message. With that in mind, I then created a switch statement that can order decoding based on one of the 5 unique ways data is stored. I couldn't think of any logic that could auto-detect which of the 5 decoding methods were used so I just used a parameter for the function to tell the function which decoding method is needed.

I also used a printf statement to display both 0x0CFFF048 and 0x0CFFF548 on the serial monitor. This is a simple way to show that I'm able to decode data correctly and able to filter more than one ID. For 0X0CFFF048 I utilized ChatGPT to create a single simulated data Frame and sent that data frame to the CAN bus using an Arduino Uno. The data organization in 0x0CFFF048 is referred to in my code as the first unique data organization. This means that there are 4 unique data sets each having a length of 2 bytes. Bytes 1 and 2 are the RPM, bytes 3 and 4 are TPS, bytes 5 and 6 are the Fuel Open Time, and bytes 7 and 8 are the Ignition angle. The simulated value sent under the 0x0CFFF048 ID is [0x20 0x1c 0x2B 0x2 0x9B 0x00 0x9C 0xFF]. Which should read RPM = 7200, TPS = 55.5%, Fuel Open Time = 15.5ms, and Ignition angle = -10 degrees. Luckily when printing the data to the serial monitor, I managed to decode these exact values for this ID. The second ID (0x0CFFF548) is also included because previously we decoded this data frame in person last time and had the data for the battery voltage. This means I can confirm my parsing is correct since I know what the battery voltage value should be.

WilliamLim781 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -117,22 +116,265 @@ void cppMain() {
+
+	/*
+	 *
+	 the table below demonstrates the organization of data in each data type. This is specifically used for the canDecodeDataFrame() in which the data type of the particular
+	 id you are decoding is required.
+
+	 	 Data type | DLC | Data Breakdown | Length			data type | DLC | Data breakdown | length			Data type | DLC | Data Breakdown | Length
+	 	 -------------------------------------------	    -------------------------------------------			-------------------------------------------
+	 	 	 1	   |  8  |     1 - 2      | 2 bytes				 3	  |  8  | 	     1       | 1 byte			     4    |  7  |	  1 - 2 	 | 2 bytes
+	  	  	  	   |	 |     3 - 4 	  | 2 bytes				      |     |        2		 | 1 byte					  |     |     3 - 4		 | 2 bytes
+	  	  	  	   |	 |     5 - 6 	  | 2 bytes				      |     |        3		 | 1 byte					  |     |     5 - 6		 | 2 bytes
+	  	  	  	   |	 |     7 - 8 	  | 2 bytes				      |     |        4		 | 1 byte					  |     |     	7	 	 | 1 bytes
+	  	 --------------------------------------------			      |     |        5		 | 1 byte			--------------------------------------------
+	  	 	 2	   |  6  |     1 - 2      | 2 bytes				      |     |        6		 | 1 byte				 5    |  8  | 	  1 - 2      | 2 bytes
+	  	  	  	   |	 |     3 - 4 	  | 2 bytes				      |     |        7		 | 1 byte					  |     |     3 - 4      | 2 bytes
+	  	  	  	   |	 |     5 - 6 	  | 2 bytes				      |     |        8		 | 1 byte					  |		|       5        | 1 byte
+	  	 --------------------------------------------		------------------------------------------					  |		|		6		 | 1 byte
+																														  |		|       7 		 | 1 byte
```
![CAN data ](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/be868d8e-a2cf-4b15-936a-0a3bcebb70f2)


The formatting be kinda off when I did the commit so I just have a snippet of the code comment here. I wanted to keep this comment here since I understand that the data type I created is kinda confusing so I wanted some visual on the code to help make sense of it

WilliamLim781 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -117,22 +116,265 @@ void cppMain() {
+				printf("\t\t 0x0CFFF548 \n");
+				printf("\r");
+				printf("Battery Volt = %.2f ", DecodeDataFrame[0]);
+				printf("Air Temp = %.2f ", DecodeDataFrame[2]);
```
![can snippet ](https://github.com/DallasFormulaRacing/DataAcquisition2.0/assets/113472670/e4337fa6-85c2-41f2-adad-89e4ce18f091)

Here is what the serial monitor looked like. The data shown on the serial monitor seems to support my testing but obviously in person testing with the ECU is needed to fully validate if the CAN parsing is correct

WilliamLim781 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -41,61 +37,64 @@ uint32_t canID;
+	canFilterSetup(9,CanIDList[8]);
+	canFilterSetup(10,CanIDList[9]);
+	canFilterSetup(11,CanIDList[10]);
+	canFilterSetup(12,CanIDList[11]);
```
I configured all the filters from the ECU for demonstration purposes. Obviously, we don't need all the filters here but I just wanted to show that the code could handle all the filters that  would be needed if we wanted to push the ECU's CAN bus to its limits

WilliamLim781 commented on Project/DAQ_System/Program/Src/app.cpp:
```diff
@@ -117,22 +116,265 @@ void cppMain() {
+	// Filter initialization
+	CAN.FilterIdHigh = FilterExtID >> 13 ;
+	CAN.FilterIdLow = ((FilterExtID & 0x1FFF) <<3) | CAN_ID_EXT | CAN_RTR_DATA ;
+	CAN.FilterMaskIdHigh = 0xFFFF;
```
in the future i would like to actually configure a proper filter mask. It would be something like 0x00001100 Where the 1 represent hex digits that do get compared and the 0s are the hex digits that are don't cares

WilliamLim781 commented:
## Future considerations to the code
1.)The CAN bus code should really be put into another folder and turned into the library or something. The CAN code adds a lot of lines to the code and makes the overall app.cpp file cluttered and hard to read.

2.) Maybe CANIDList array could be turned into a #define. That may save space compared to using the array.

3.) For canFilterSetup() It might be a good idea to just automatically configure all filters or have two separate functions that select which id to filter and another function that just activates all filters.

4.) I also realized that I forgot to add the message flags properly =(. The flag is set in  the callback function but never reset so it kinda is true for the whole duration of the code.

5.) An overhaul on the data types of the canDecodeDataFrame() would also be nice. It works and it allows me to reduce the amount of lines needed to decode all the functions but I don't think it's very friendly to others reading the code. I think the approach is ok but the idea itself is kinda confusing or hard to convey via the comments.

crsz20 commented:
Great observation on defining your own generalized set of data types! I have some design ideas for improvement:

- You are defining custom data types and assigning numbers to them as an indicator. C++ [enums](https://www.programiz.com/cpp-programming/enumeration) allows you to do this while declaring a name for each. Suppose we name this enum `PE_Message`

```c++
enum PE_Message_Classification {
    FORMAT1,
    FORMAT2,
    ...
};
```
- We can define a lookup table to associate a CAN ID with one of the formats of the enum.
- Then, each enum format can have a corresponding struct.

```c++
struct Format4_Messagge {
   std::pair <uint8_t, uint8_t> field1; // 2 bytes
   std::pair <uint8_t, uint8_t> field2;
   std::pair <uint8_t, uint8_t> field3;
   std::pair <uint8_t, uint8_t field4; // Just 1 byte
}
```

- So, in practice, something along the lines of...

Pseudocode:
```
Decode() {
	PE_Message_Classification message_classification = lookup_table.val(canID)
	
	switch(message_classification)
           ...
           case(FORMAT4):  Format4_Messagge = Parse_Format_1(message_buffer, canID)
           ...

	switch(canID)
            ...
           case(0x0CFFF148): 
		RPM = Format4_Message.field1
		TPS = Format4_Message.field2
}

```

The constructor of each Message class can use parse 1 or 2 bytes at a time, accordingly.

crsz20 commented:
I think I have a more fleshed-out idea that's more reflective of the ECU's datasheet.

* The conversion algorithm can be a function: `int ParseBytePair(low_byte, high_byte)`.
* I will refer to your custom data type that you enumerated 1 through 5 as "Format1", "Format2", etc.
* By further examining the datasheet:
  * Frame "PE1" **is a** Format1 message.
  * Frame "PE2" **is a** Format4 message.
  * Notice the "is a" relationship — this is a potential use case for inheritance.
 
Consider each format as a structure:
```c++
struct Format1 {
  int num_of_fields = 4;
  int[num_of_fields] fields;

  // Constructor
  Format1(RxData[]) {
    // Parse RxData array using the BytePairParse function, storing the results in the fields[ ] array
  } 
};
```

This format will be re-used across different specific PE frames. So, let's define a class for the frame that inherits the format structure:
```c++
class PE1_Frame : public Format1 {
  // Constructor calls the constructor of Format1 to get access to fields
 PE1_Frame (RxData []) { ... }
  
  rpm_resolution_per_bit = 1;
  tps_resolution_per_bit = 0.1;

  GetRPM() { return fields[1] * rpm_resolution_per_bit; }
  GetTPS() { return fields[2] * tps_resolution_per_bit; }
}
```

* So, in practice, decoding upon receiving a new CAN message will look like this:
  * Identify CAN ID
  * Create the appropriate `PEx_Frame`, passing `RxData` to its constructor at initialization
  * Access final computed data using `PEx_Frame.GetRPM( )`
  * Pass the final computed data to the ECU class holding **all** data (we can then do input validation here)

Pseudocode:
```
Decode() {
  switch(canID) {
  case(0x0CFFF048):
    PE1_Frame frame(RxData[]);
    float rpm = frame.GetRPM();
    ...
  }
  case(...)
}

WilliamLim781 committed:
can interrupt flag bug fix

Previously I forgot to reset the flag bit to 0 after each iteration. So at the end of the if statement that checks for the flag bit status, I reset the flag bit
